hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/NativeSparkDataSet.java

```diff
@@ -0,0 +1,1048 @@
+            }
+        });
+    }
+
+    @Override
+    public <Op extends SpliceOperation, K> PairDataSet< K, V> keyBy(SpliceFunction<Op, V, K> f) {
+        try {
+            return new SparkDataSet<>(NativeSparkDataSet.<V>toSpliceLocatedRow(dataset, this.context)).keyBy(f);
+        } catch (Exception e) {
+            throw Exceptions.throwAsRuntime(e);
+        }
+    }
+
+    public <Op extends SpliceOperation, K> PairDataSet< K, V> keyBy(SpliceFunction<Op, V, K> f, String name) {
+        return keyBy(f);
+    }
+
+
+    public <Op extends SpliceOperation, K> PairDataSet< K, V> keyBy(SpliceFunction<Op, V, K> f, OperationContext context) throws StandardException {
+        return new SparkPairDataSet<>(NativeSparkDataSet.<V>toSpliceLocatedRow(dataset, this.context).keyBy(new SparkSpliceFunctionWrapper<>(f)), f.getSparkName());
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    @Override
+    public <Op extends SpliceOperation, K> PairDataSet< K, V> keyBy(
+            SpliceFunction<Op, V, K> f, String name, boolean pushScope, String scopeDetail) {
+        return keyBy(f);
+    }
+
+    @Override
+    public long count() {
+        return dataset.count();
+    }
+
+    @Override
+    public DataSet<V> union(DataSet<V> dataSet, OperationContext operationContext) {
+        return union(dataSet, operationContext, RDDName.UNION.displayName(), false, null);
+    }
+
+
+    @Override
+    public DataSet<V> parallelProbe(List<ScanSetBuilder<ExecRow>> scanSetBuilders, OperationContext<MultiProbeTableScanOperation> operationContext) throws StandardException {
+        DataSet<V> toReturn = null;
+        DataSet<V> branch = null;
+        int i = 0;
+        MultiProbeTableScanOperation operation = operationContext.getOperation();
+        for (ScanSetBuilder<ExecRow> builder: scanSetBuilders) {
+            DataSet<V> dataSet = (DataSet<V>) builder.buildDataSet(operation);
+            if (i % 100 == 0) {
+                if (branch != null) {
+                    if (toReturn == null)
+                        toReturn = branch;
+                    else
+                        toReturn = toReturn.union(branch, operationContext);
+                }
+                branch = dataSet;
+            }
+            else
+                branch = branch.union(dataSet, operationContext);
+            i++;
+        }
+        if (branch != null) {
+            if (toReturn == null)
+                toReturn = branch;
+            else
+                toReturn = toReturn.union(branch, operationContext);
+        }
+        return toReturn;
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    @Override
+    public DataSet< V> union(DataSet<V> dataSet, OperationContext operationContext, String name, boolean pushScope, String scopeDetail) {
+        pushScopeIfNeeded((SpliceFunction)null, pushScope, scopeDetail);
+        try {
+            Dataset rdd1 = dataset.union(((NativeSparkDataSet) dataSet).dataset);
```

Comment:
dataset could be a SparkDataSet instead

Reply:
You may want to convert to NativeDataSet, similar to intersect?